types: use AdjustedGoTime in ConvertTimeZone to fix DST conversion error

minleejae opened the pull request:
### What problem does this PR solve?

Issue Number: close #65299

Problem Summary:

When `SELECT DISTINCT` is used with both datetime and timestamp columns, the query can fail with "Incorrect datetime value" error due to timezone conversion issues in `ConvertTimeZone()` during DST transitions.

### What changed and how does it work?

Changed `ConvertTimeZone()` to use `AdjustedGoTime()` instead of `GoTime()`.

The issue occurs because `GoTime()` fails validation for times that don't exist during DST transitions (e.g., 2:30 AM when clocks jump from 2:00 AM to 3:00 AM). `AdjustedGoTime()` handles these cases by adjusting to the nearest valid timezone boundary.

### Check List

Tests <!-- At least one of them must be included. -->

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
Fix timezone conversion error in SELECT DISTINCT with datetime and timestamp columns during DST transitions
```

minleejae committed:
types: use AdjustedGoTime in ConvertTimeZone to fix DST conversion error

When SELECT DISTINCT is used with both datetime and timestamp columns,
the query can fail with "Incorrect datetime value" error due to timezone
conversion issues in ConvertTimeZone() during DST transitions.

The issue occurs because GoTime() fails validation for times that don't
exist during DST transitions (e.g., 2:30 AM when clocks jump from 2:00 AM
to 3:00 AM). AdjustedGoTime() handles these cases by adjusting to the
nearest valid timezone boundary.

lance6716 commented:
CI failed

```
run test [types/time] err: sql:SELECT *, UNIX_TIMESTAMP(ts) FROM t WHERE ts < '2025-03-30 02:00:00';: failed to run query
"SELECT *, UNIX_TIMESTAMP(ts) FROM t WHERE ts < '2025-03-30 02:00:00';"
 around line 161,
 we need(128):
Level	Code	Message
Warning	8179	Timestamp is not valid, since it is in Daylight Saving Time transition '{2025 3 30 2 0 0 0}' for
but got(128):
SELECT *, UNIX_TIMESTAMP(ts) FROM t WHERE ts < '2025-03-30 02:00:00';
id	ts	UNIX_TIMESTAMP(ts)
1	2025-03-30 01:59:59	1743296399
```

please check if it's expected

minleejae committed:
types: fix DST conversion while preserving warning

When ConvertTimeZone encounters a DST transition time that doesn't exist,
first try GoTime() and if it fails, use AdjustedGoTime() to adjust the time
to the nearest valid time. Return the original error so callers can handle
it as a warning, maintaining backward compatibility.

This fixes the bug where SELECT DISTINCT with datetime/timestamp columns
would fail during DST transitions, while still showing the appropriate
warning message to users.

lance6716 commented:
/cc @mjonss 
Can you also help to triage the issue
